src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Cache.cs

```diff
@@ -134,6 +134,8 @@ public static JsonPropertyInfo AddProperty(Type propertyType, PropertyInfo prope
             JsonConverter converter,
             JsonSerializerOptions options)
         {
+            Debug.Assert(!(IsInvalidForSerialization(declaredPropertyType) || IsInvalidForSerialization(runtimePropertyType)));
```

Comment:
Is there coverage for `runtimePropertyType`? I was expecting to see test classes with a property and\or root `object` Type (which uses polymorphic behavior and thus uses `runtimePropertyType` during serialization).

Reply:
Looks like none of the types we are guarding against - pointers, ref structs, open generics - can have instances that can be boxed. Pointers & ref structs can't be made to derive from anything else AFAIK.

An instance can only have a declared type which is an open generic at the root API level, and we have tests that guard against that for both serialization & deserialization.

Reply:
So do we need to have the condition then for `runtimePropertyType`?

Reply:
Given this closed set of invalid type categories we are checking for, I can't think of a scenario where the declared type (validated upstream) is valid but the runtime type is not. This may change as/if we discover more invalid types, but for now, I think the `Debug.Assert` added in https://github.com/dotnet/runtime/pull/34777/commits/a078047d8ad29ddb9611297d6354db996a51eee1 suffices.